cpp/src/arrow/dataset/filter.cc

```diff
@@ -220,6 +221,23 @@ std::shared_ptr<Expression> Invert(const Expression& expr) {
   return nullptr;
 }
 
+std::shared_ptr<Expression> Expression::Assume(const Expression& given) const {
+  if (given.type() == ExpressionType::COMPARISON) {
+    const auto& cmp = checked_cast<const ComparisonExpression&>(given);
+    if (cmp.op() == compute::CompareOperator::EQUAL) {
+      if (this->Equals(cmp.left_operand())) {
+        return cmp.right_operand();
```

Comment:
I'm not sure of the point of this. It seems to be assuming that `right_operand` is somehow simpler (and therefore more desirable) than `this`. But is it the case generally?

Perhaps this should be restricted to constant comparands...

Reply:
The point is to allow `"beta"_.Assume("beta"_ == 3).Equals(3)`,
which in turn allows `"beta"_.In({1, 2, 3}).Assume("beta"_ == 0).Equals(false)`.

Restricting it to constant given.right_operand seems sensible